### ISSUE TITLE
Flaky crash of `samples/ffi/async/async_test` 

### ISSUE THREAD

**Issue by mkustermann** [dart-lang/sdk#49361]
From [this log](https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket/8810140386405595761/+/u/test_results/ignored_flaky_test_failure_logs):
```

/===================================================================\
| samples/ffi/async/async_test broke (Pass -> Crash, expected Pass) |
\===================================================================/

--- Command "vm" (took 445ms):
DART_CONFIGURATION=ReleaseX64 xcodebuild/ReleaseX64/dart --sound-null-safety -Dtest_runner.configuration=dartk-strong-mac-release-x64 --ignore-unrecognized-flags --packages=/opt/s/w/ir/.dart_tool/package_config.json /opt/s/w/ir/samples/ffi/async/async_test.dart

exit code:
-6

stdout:
Dart = Dart mutator thread executing Dart.
C Da = Dart mutator thread executing C.
C T1 = Some C thread executing C.
C T2 = Some C thread executing C.
C    = C T1 or C T2.
Dart: Setup.
Dart: Tell C to start worker threads.
C Da:  Starting SimulateWork.
C Da:   Starting worker threads.
C Da:  Started SimulateWork.
Dart: Yielding (able to receive messages on port).
C T1: Work1 Start.
C T2: Work2 Start.
C T2: MyCallbackNonBlocking(5).
C   :  Posting message (port: cfee9377cb95f, work: 7f81a0204420).
C T2: MyCallbackBlocking(1).
C   :  Posting message (port: cfee9377cb95f, work: 7f81a0204e40).
C   :  Waiting for result.
Dart = Dart mutator thread executing Dart.
C Da = Dart mutator thread executing C.
C T1 = Some C thread executing C.
C T2 = Some C thread executing C.
C    = C T1 or C T2.
Dart: Setup.
Dart: Tell C to start worker threads.
C Da:  Starting SimulateWork.
C Da:   Starting worker threads.
C Da:  Started SimulateWork.
C T1: Work1 Start.
Dart: Yielding (able to receive messages on port).
C T2: Work2 Start.
C T2: MyCallback2(5).
C   :  Dart_PostCObject_(request: 700009b8fee0, call: 700009b8fe50).
C T2: MyCallback1(1).
C   :  Dart_PostCObject_(request: 700009b8fd00, call: 700009b8fc70).
C   :  Waiting for result.
Dart:   Calling into C to execute callback (Pointer<Work>: address=0x7f81a0204420).
C Da:    ExecuteCallback(00007f81a0204420).
Dart:     callback2(5).
C Da:    ExecuteCallback done.
Dart:   Done with callback.
Dart:   Calling into C to execute callback (Pointer<Work>: address=0x7f81a0204e40).
C Da:    ExecuteCallback(00007f81a0204e40).
Dart:     callback1(1).
C Da:     Notify result ready.
C   :  Received result.
C T2: MyCallbackBlocking returned 4.
C T2: MyCallbackNonBlocking(4).
C   :  Posting message (port: cfee9377cb95f, work: 7f819e60d820).
C T2: Work2 Done.

stderr:
===== CRASH =====
si_signo=Segmentation fault: 11(11), si_code=1, si_addr=0x700009a89eef
version=2.18.0-edge.e87c84ff699400beb7232a256ce368b42b2394c0 (be) (Tue Jun 28 03:41:12 2022 +0000) on "macos_x64"
pid=22893, thread=7939, isolate_group=main(0x7f81a2008600), isolate=main(0x7f81a200e200)
isolate_instructions=10f023580, vm_instructions=10f023580
  pc 0x00007fff68fb86da fp 0x0000700009983220 _pthread_cond_updateval+0x3e
  pc 0x00007fff68fba540 fp 0x0000700009983280 pthread_cond_signal+0x209
  pc 0x000000011cd344e9 fp 0x0000700009983290 std::__2::condition_variable::notify_one()+0x9
  pc 0x000000011cde9b1c fp 0x00007000099832e0 ExecuteCallback+0x5c
  pc 0x0000000111d05bbb fp 0x0000700009983308 Unknown symbol
  pc 0x00000001156390b3 fp 0x0000700009983338 Unknown symbol
  pc 0x00000001156380ae fp 0x0000700009983388 Unknown symbol
  pc 0x0000000115637edc fp 0x00007000099833b8 Unknown symbol
  pc 0x0000000115637d51 fp 0x0000700009983438 Unknown symbol
  pc 0x00000001156379b8 fp 0x0000700009983490 Unknown symbol
  pc 0x00000001156375f3 fp 0x00007000099834d0 Unknown symbol
  pc 0x0000000115637088 fp 0x0000700009983510 Unknown symbol
  pc 0x0000000115636cce fp 0x0000700009983550 Unknown symbol
  pc 0x00000001156369cf fp 0x0000700009983590 Unknown symbol
  pc 0x00000001156367cd fp 0x00007000099835d0 Unknown symbol
  pc 0x0000000115622fcf fp 0x0000700009983638 Unknown symbol
  pc 0x0000000115622843 fp 0x0000700009983688 Unknown symbol
  pc 0x0000000111d02a0c fp 0x0000700009983700 Unknown symbol
  pc 0x000000010f180a4c fp 0x00007000099837a0 dart::DartEntry::InvokeCode(dart::Code const&, unsigned long, dart::Array const&, dart::Array const&, dart::Thread*)+0x14c
  pc 0x000000010f1808ab fp 0x0000700009983800 dart::DartEntry::InvokeFunction(dart::Function const&, dart::Array const&, dart::Array const&, unsigned long)+0x12b
  pc 0x000000010f182ae2 fp 0x0000700009983850 dart::DartLibraryCalls::HandleMessage(long long, dart::Instance const&)+0x132
  pc 0x000000010f1ab045 fp 0x0000700009983df0 dart::IsolateMessageHandler::HandleMessage(std::__2::unique_ptr<dart::Message, std::__2::default_delete<dart::Message>>)+0x365
  pc 0x000000010f1d6c0d fp 0x0000700009983e60 dart::MessageHandler::HandleMessages(dart::MonitorLocker*, bool, bool)+0x12d
  pc 0x000000010f1d731b fp 0x0000700009983eb0 dart::MessageHandler::TaskCallback()+0x1db
  pc 0x000000010f30d8f7 fp 0x0000700009983f30 dart::ThreadPool::WorkerLoop(dart::ThreadPool::Worker*)+0x147
  pc 0x000000010f30dd69 fp 0x0000700009983f60 dart::ThreadPool::Worker::Main(unsigned long)+0x79
  pc 0x000000010f27921f fp 0x0000700009983fb0 dart::ThreadStart(void*)+0x9f
  pc 0x00007fff68fbb109 fp 0x0000700009983fd0 _pthread_start+0x94
  pc 0x00007fff68fb6b8b fp 0x0000700009983ff0 thread_start+0xf
-- End of DumpStackTrace
  pc 0x0000000000000000 fp 0x0000700009983308 sp 0x0000000000000000 [Optimized] 2296659968
  pc 0x00000001156390b3 fp 0x0000700009983338 sp 0x0000700009983318 [Optimized] FfiTrampoline_executeCallback
  pc 0x00000001156380ae fp 0x0000700009983388 sp 0x0000700009983348 [Unoptimized] requestExecuteCallback
  pc 0x0000000115637edc fp 0x00007000099833b8 sp 0x0000700009983398 [Unoptimized] requestExecuteCallback
  pc 0x0000000115637d51 fp 0x0000700009983438 sp 0x00007000099833c8 [Unoptimized] _RootZone@4048458.runUnaryGuarded
  pc 0x00000001156379b8 fp 0x0000700009983490 sp 0x0000700009983448 [Unoptimized] _BufferingStreamSubscription@4048458._sendData@4048458
  pc 0x00000001156375f3 fp 0x00007000099834d0 sp 0x00007000099834a0 [Unoptimized] _BufferingStreamSubscription@4048458._add@4048458
  pc 0x0000000115637088 fp 0x0000700009983510 sp 0x00007000099834e0 [Unoptimized] _SyncStreamController@4048458._sendData@4048458
  pc 0x0000000115636cce fp 0x0000700009983550 sp 0x0000700009983520 [Unoptimized] _StreamController@4048458._add@4048458
  pc 0x00000001156369cf fp 0x0000700009983590 sp 0x0000700009983560 [Unoptimized] _StreamController@4048458.add
  pc 0x00000001156367cd fp 0x00007000099835d0 sp 0x00007000099835a0 [Unoptimized] _StreamController@4048458.add
  pc 0x0000000115622fcf fp 0x0000700009983638 sp 0x00007000099835e0 [Unoptimized] _Closure@0150898.dyn:call
  pc 0x0000000115622843 fp 0x0000700009983688 sp 0x0000700009983648 [Unoptimized] _RawReceivePortImpl@1026248._handleMessage@1026248
  pc 0x0000000111d02a0c fp 0x0000700009983700 sp 0x0000700009983698 [Stub] InvokeDartCode

--- Re-run this test:
python3 tools/test.py -n dartk-strong-mac-release-x64 samples/ffi/async/async_test
```

/cc @dcharkes

**Comment by dcharkes** [comment]
Looks like this test has been flakily crashing and timing out for a while ([flakiness dashboard](https://dashboards.corp.google.com/_0d7e1e34_2c58_4d29_bf12_cf7bed10396d?fb=name:ct:ffi%2Fasync%2Fasync_test)).

Example timeout log

```
stdout:
Dart = Dart mutator thread executing Dart.
C Da = Dart mutator thread executing C.
C T1 = Some C thread executing C.
C T2 = Some C thread executing C.
C    = C T1 or C T2.
Dart: Setup.
Dart: Tell C to start worker threads.
C Da:  Starting SimulateWork.
C Da:   Starting worker threads.
C Da:  Started SimulateWork.
Dart: Yielding (able to receive messages on port).
C T1: Work1 Start.
Dart = Dart mutator thread executing Dart.
C Da = Dart mutator thread executing C.
C T1 = Some C thread executing C.
C T2 = Some C thread executing C.
C    = C T1 or C T2.
Dart: Setup.
Dart: Tell C to start worker threads.
C Da:  Starting SimulateWork.
C Da:   Starting worker threads.
C T2: Work2 Start.
C Da:  Started SimulateWork.
C T2: MyCallbackNonBlocking(5).
C   :  Posting message (port: b2824ca22a8eb, work: 12a10aef0).
C T1: Work1 Start.
C T2: Work2 Start.
C T2: MyCallback2(5).
C   :  Dart_PostCObject_(request: 16dfc2e30, call: 16dfc2ec0).
C T2: MyCallback1(1).
C   :  Dart_PostCObject_(request: 16dfc2c28, call: 16dfc2cb8).
Dart: Yielding (able to receive messages on port).
Dart:   Got message: CppRequest(method: myCallback2, 1 bytes)
Dart:     myCallback2(5).
Dart:   Got message: CppRequest(method: myCallback1, 1 bytes)
Dart:     myCallback1(1).
Dart:   Responding: CppResponse(message: 1)
C T2: MyCallbackBlocking(1).
C   :  Posting message (port: b2824ca22a8eb, work: 12970de90).
C   :   HandleResponse (call: 103da4048, message: 103da4128).
C   :    ResolveCall(length: 1, buffer: 12a007c25).
C   :     Notify result ready.
C   :  Waiting for result.
C   :  Received result.
C T2: MyCallback1 returned 4.
C T2: MyCallback2(4).
C   :  Dart_PostCObject_(request: 16dfc2e30, call: 16dfc2ec0).
C T2: Work2 Done.
Dart:   Calling into C to execute callback (Pointer<Work>: address=0x12a10aef0).
C Da:    ExecuteCallback(000000012a10aef0).
Dart:     callback2(5).
C Da:    ExecuteCallback done.
Dart:   Done with callback.
Dart:   Calling into C to execute callback (Pointer<Work>: address=0x12970de90).
C Da:    ExecuteCallback(000000012970de90).
C   :  Waiting for result.
Dart:     callback1(1).
C Da:     Notify result ready.
C Da:    ExecuteCallback done.
Dart:   Done with callback.
Dart:   Got message: CppRequest(method: myCallback2, 1 bytes)
Dart:     myCallback2(4).
Dart: Yielding (able to receive messages on port).
Dart: Yielding (able to receive messages on port).
C T1: MyCallback1(3).
C   :  Dart_PostCObject_(request: 16df36c58, call: 16df36ce8).
C   :  Waiting for result.
Dart:   Got message: CppRequest(method: myCallback1, 1 bytes)
Dart:     myCallback1(3).
Dart:   Responding: CppResponse(message: 1)
C   :   HandleResponse (call: 103da4048, message: 103da4128).
C   :    ResolveCall(length: 1, buffer: 12a10a895).
C   :     Notify result ready.
C   :  Received result.
C T1: MyCallback1 returned 6.
C T1: MyCallbackBlocking(3).
C   :  Posting message (port: b2824ca22a8eb, work: 12a108a60).
C   :  Waiting for result.
Dart:   Calling into C to execute callback (Pointer<Work>: address=0x12a108a60).
C Da:    ExecuteCallback(000000012a108a60).
Dart:     callback1(3).
C Da:     Notify result ready.
C Da:    ExecuteCallback done.
Dart:   Done with callback.
C   :  Received result.
C T1: MyCallbackBlocking returned 6.
Dart: Yielding (able to receive messages on port).
Dart: Yielding (able to receive messages on port).
Dart: Yielding (able to receive messages on port).
Dart: Yielding (able to receive messages on port).
C T1: MyCallback2(5).
C   :  Dart_PostCObject_(request: 16df36e68, call: 16df36ef8).
C T1: Work1 Done.
Dart:   Got message: CppRequest(method: myCallback2, 1 bytes)
Dart:     myCallback2(5).
C T1: MyCallbackNonBlocking(5).
C   :  Posting message (port: b2824ca22a8eb, work: 11970ad80).
C T1: Work1 Done.
Dart:   Calling into C to execute callback (Pointer<Work>: address=0x11970ad80).
C Da:    ExecuteCallback(000000011970ad80).
Dart:     callback2(5).
C Da:    ExecuteCallback done.
Dart:   Done with callback.
Dart: Yielding (able to receive messages on port).
Dart: Received expected number of callbacks.
Dart: Tell C to stop worker threads.
C Da:  Stopping SimulateWork.
C Da:   Waiting for worker threads to finish.
C Da:  Stopped SimulateWork.
Dart: Done.
Dart: Yielding (able to receive messages on port).
Dart: Yielding (able to receive messages on port).
Dart: Yielding (able to receive messages on port).
Dart: Yielding (able to receive messages on port).
Dart: Yielding (able to receive messages on port).
Dart: Yielding (able to receive messages on port).
Dart: Yielding (able to receive messages on port).
...
```

[full log](https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket/8811723200157739889/+/u/test_results/ignored_flaky_test_failure_logs).

I'll investigate.

**Comment by Cat-sushi** [comment]
Stable release of 2.18.0 also have a similar bug.
```
===== CRASH =====
si_signo=Segmentation fault(11), si_code=1, si_addr=0x29
version=2.18.0 (stable) (Unknown timestamp) on "linux_x64"
pid=352686, thread=352794, isolate_group=main(0x560573359800), isolate=Server.main(0x560573812800)
isolate_instructions=56057192d440, vm_instructions=56057192d440
  pc 0x00007f7529dad6b0 fp 0x00007f7519e7da58 Unknown symbol
  pc 0x00007f7521791c2a fp 0x00007f7519e7dac0 Unknown symbol
  pc 0x00007f7529db7ca4 fp 0x00007f7519e7db08 Unknown symbol
  pc 0x00007f75217ead68 fp 0x00007f7519e7db50 Unknown symbol
  pc 0x00007f75217eac56 fp 0x00007f7519e7db90 Unknown symbol
  pc 0x00007f75217d556e fp 0x00007f7519e7dbc8 Unknown symbol
  pc 0x00007f75217d5334 fp 0x00007f7519e7dc28 Unknown symbol
  pc 0x00007f75217eabb6 fp 0x00007f7519e7dc90 Unknown symbol
  pc 0x00007f7515b59a05 fp 0x00007f7519e7dce8 Unknown symbol
  pc 0x00007f751708effc fp 0x00007f7519e7dd68 Unknown symbol
  pc 0x00007f7529dbfd37 fp 0x00007f7519e7de20 Unknown symbol
  pc 0x00007f7529db90d5 fp 0x00007f7519e7deb8 Unknown symbol
  pc 0x00007f752726217d fp 0x00007f7519e7df08 Unknown symbol
  pc 0x00007f752725113c fp 0x00007f7519e7df60 Unknown symbol
  pc 0x00007f7527250d9e fp 0x00007f7519e7dfb0 Unknown symbol
  pc 0x00007f7527250abe fp 0x00007f7519e7e010 Unknown symbol
  pc 0x00007f752724fdc7 fp 0x00007f7519e7e090 Unknown symbol
  pc 0x00007f7527261f2a fp 0x00007f7519e7e0d8 Unknown symbol
  pc 0x00007f75272618c1 fp 0x00007f7519e7e130 Unknown symbol
  pc 0x00007f75272615fd fp 0x00007f7519e7e170 Unknown symbol
  pc 0x00007f752725aa71 fp 0x00007f7519e7e1f0 Unknown symbol
  pc 0x00007f752725a540 fp 0x00007f7519e7e248 Unknown symbol
  pc 0x00007f7529dbf294 fp 0x00007f7519e7e290 Unknown symbol
  pc 0x00007f7527259bb8 fp 0x00007f7519e7e2d0 Unknown symbol
  pc 0x00007f752725990e fp 0x00007f7519e7e310 Unknown symbol
  pc 0x00007f752725960f fp 0x00007f7519e7e350 Unknown symbol
  pc 0x00007f75217e5fbd fp 0x00007f7519e7e390 Unknown symbol
  pc 0x00007f7527223fbf fp 0x00007f7519e7e3f8 Unknown symbol
  pc 0x00007f7527223833 fp 0x00007f7519e7e448 Unknown symbol
  pc 0x00007f752e202a0c fp 0x00007f7519e7e4c0 Unknown symbol
  pc 0x0000560571aa55fd fp 0x00007f7519e7e560 dart::DartEntry::InvokeCode(dart::Code const&, unsigned long, dart::Array const&, dart::Array const&, dart::Thread*)+0x14d
  pc 0x0000560571aa543c fp 0x00007f7519e7e5c0 dart::DartEntry::InvokeFunction(dart::Function const&, dart::Array const&, dart::Array const&, unsigned long)+0x14c
  pc 0x0000560571aa786c fp 0x00007f7519e7e610 dart::DartLibraryCalls::HandleMessage(long, dart::Instance const&)+0x14c

  pc 0x0000560571aceee0 fp 0x00007f7519e7eba0 dart::IsolateMessageHandler::HandleMessage(std::__2::unique_ptr<dart::Message, std::__2::default_delete<dart::Message> >)+0x350
  pc 0x0000560571af840d fp 0x00007f7519e7ec10 dart::MessageHandler::HandleMessages(dart::MonitorLocker*, bool, bool)+0x14d
  pc 0x0000560571af8aef fp 0x00007f7519e7ec60 dart::MessageHandler::TaskCallback()+0x1df
  pc 0x0000560571c1b9a8 fp 0x00007f7519e7ece0 dart::ThreadPool::WorkerLoop(dart::ThreadPool::Worker*)+0x148
  pc 0x0000560571c1bdfd fp 0x00007f7519e7ed10 dart::ThreadPool::Worker::Main(unsigned long)+0x6d
  pc 0x0000560571b8e5c8 fp 0x00007f7519e7edd0 /usr/lib/dart/bin/dart+0x21615c8
-- End of DumpStackTrace

Aborted (core dumped)
Exited (134)
```

**Comment by rmacnak-google** [comment]
The flaky dashboard shows this as active=false. Probably fixed by ed5aeab889ef52e86cec36318e9cd8c2b796dccf and 06dd95bd8e5f12ffcf031d86fa9905b7d95ff78d.
